tutorials/model_inference.md

```diff
@@ -0,0 +1,159 @@
+
+The 223 ontology contains the rules we will use for inference.
+We will load the 223 ontology into a separate graph from the model.
+This is mostly for maintenance: it is easy enough to merge two graphs together into one, but it is much harder to factor them out again.
+By keeping the ontology graph separate from the model graph, we can more easily maintain and version those graphs individually.
+
+```{code-cell}
+from rdflib import Graph
+# Create a Graph
+s223 = Graph()
+# Parse in a recent copy of the 223 ontology
+s223.parse("https://query.open223.info/ontologies/223p.ttl")
+```
+
+## Applying Inference Rules
+
+To apply inference rules, we need an *inference engine*.
+This is a piece of software which knows how to properly interpret and apply the SHACL rules defined in an ontology.
+There are multiple closed-source and open-source implementations of SHACL available; some of these are listed on [this page](other-resources)
+Below, we will be using the open-source [PySHACL](https://github.com/RDFLib/pySHACL) library.
+
+We import the PySHACL library and then invoke the `validate` function on our model graph (*data graph* in PySHACL parlance)
+and our 223 graph (*shape graph* in PySHACL parlance).
+
+```{code-cell}
+import pyshacl
+
+# skolemizing the s223 graph lets us remove blank nodes after inference
+skolemized_s223 = s223.skolemize()
+
+pyshacl.validate(model,
+    shacl_graph=skolemized_s223,     # pass in the 223 graph object here
+    ont_graph=skolemized_s223,       # pass in the 223 graph object here
+    allow_infos=True,     # don't fail if we get an INFO message
+    allow_warnings=True,  # don't fail if we get a WARNING message
+    abort_on_first=False, # allow errors to happen during execution
+    advanced=True,        # allow SHACL rules to execute
+    inplace=True          # update the 'model' object with the inferred triples
+)
+# remove the skolemized s223 graph from the model
+model -= skolemized_s223
+# de-skolemize the ret of the model
+model = model.de_skolemize()
+print(f"The model now contains {len(model)} triples")
+```
+
+This may take a few minutes to run, depending on the size of your model.
+If the PySHACL library is too slow, we recommend looking at alternate open-source implementations
+like [TopQuadrant's Java-based implementation](https://github.com/TopQuadrant/shacl).
+
+We can see from the print statement that several triples have been added to the model.
+
+```{note}
+The `pyshacl.validate` function actually returns a SHACL validation report which can be used to fix
+the model and make it compatible with the 223 ontology. See the [Model Validation](model_validation)
+tutorial for how to access and interpret this report.
+```
+
+<details>
+<summary>Turtle representation of the model (post-inference)</summary>
+
+```{code-cell}
+print(model.serialize())
+```
+
+</details>
+
+## Using the New Model
+
+To demonstrate that the model contains new triples, we can try re-running the query from before.
```

Comment:
You can also subtract the original graph from the new validated graph to show all the added triples, which may be nice.

Reply:
This is a bit more juggling than I want to do for this tutorial, but maybe we can consider showing how this works under an "explanation" page for the `cnx` relationships